src/flirtlib.cpp

```diff
@@ -39,52 +39,87 @@
+  nh_.param("descriptor_min_rho", descriptor_min_rho_, descriptor_min_rho_);
+  nh_.param("descriptor_max_rho", descriptor_max_rho_, descriptor_max_rho_);
+  nh_.param("descriptor_bin_rho", descriptor_bin_rho_, descriptor_bin_rho_);
+  nh_.param("descriptor_bin_phi", descriptor_bin_phi_, descriptor_bin_phi_);
 
-FlirtlibFeatures::FlirtlibFeatures (ros::NodeHandle nh)
-{
-  ROS_INFO("Note: currently ignoring ros params and using hardcoded params");
   peak_finder_.reset(new SimpleMinMaxPeakFinder(0.34, 0.001));
   histogram_dist_.reset(new SymmetricChi2Distance<double>());
-  detector_.reset(createDetector(peak_finder_.get()));
-  descriptor_.reset(createDescriptor(histogram_dist_.get()));
-  ransac_.reset(new RansacFeatureSetMatcher(0.0599, 0.95, 0.4, 0.4,
-                                            0.0384, false));
+
+  // Detector
+  if (detector_type_ == "curvature")
+  {
+    CurvatureDetector * det = new CurvatureDetector(peak_finder_.get(), detector_scale_, detector_base_sigma_,
+                                                    detector_sigma_step_,detector_minimum_spanning_trees_num_);
+    det->setUseMaxRange(false);
+    detector_ = boost::shared_ptr<Detector>(det);
+  }
+  else if (detector_type_ == "normal_edge")
+  {
+    detector_ .reset(new NormalEdgeDetector(peak_finder_.get(), detector_scale_, detector_base_sigma_,
+                                            detector_sigma_step_,  detector_window_));
+  }
+  else if (detector_type_ == "normal blob")
+  {
+    detector_.reset(new NormalBlobDetector(peak_finder_.get(),detector_scale_, detector_base_sigma_,
+                                           detector_sigma_step_,  detector_window_));
+  }
+  else if (detector_type_ == "range")
+  {
+    detector_.reset(new RangeDetector(peak_finder_.get(), detector_scale_, detector_base_sigma_,
+                                      detector_sigma_step_));
+  }
+
+  // Descriptor
+  if(descriptor_type_ == "Beta Grid")
+  {
+    BetaGridGenerator * gen = new BetaGridGenerator(descriptor_min_rho_, descriptor_max_rho_,
+                                                    descriptor_bin_rho_, descriptor_bin_phi_);
+    gen->setDistanceFunction(histogram_dist_.get());
+    descriptor_ .reset(gen);
+  }
+  if(descriptor_type_ == "Shape Context")
+  {
+    ShapeContextGenerator * gen = new ShapeContextGenerator(descriptor_min_rho_, descriptor_max_rho_,
+                                                            descriptor_bin_rho_, descriptor_bin_phi_);
+    gen->setDistanceFunction(histogram_dist_.get());
+    descriptor_ .reset(gen);
+  }
+
+  ransac_.reset(new RansacFeatureSetMatcher(0.0299, 0.95, 0.2, 0.1,0.0184, false));
 }
 
-// Extract flirtlib features
 InterestPointVec
-FlirtlibFeatures::extractFeatures (sm::LaserScan::ConstPtr scan) const
+FlirtlibFeatures::extractFeatures (sensor_msgs::LaserScan::ConstPtr scan) const
 {
   boost::shared_ptr<LaserReading> reading = fromRos(*scan);
```

Comment:
I guess you meant FlirtlibFeatures::extractFeatures (const sensor_msgs::LaserScan& scan) const above?